src/main/ank/2.2-limits-and-colimits.md

```diff
@@ -0,0 +1,210 @@
+```Haskell
+p :: c -> a
+q :: c -> b
+```
+```kotlin
+val p: (C) -> A
+val q: (C) -> B
+```
+................
+```Haskell
+q = f . p
+q = g . p
+```
+```kotlin
+val q = f compose p
+val q = g compose p
+```
+................
+```Haskell
+f . p = g . p
+```
+```kotlin
+f compose p == g compose p
+```
+................
+```Haskell
+f (x, y) = 2 * y + x
+g (x, y) = y - x
+```
+```kotlin:ank:playground
+val f = { x: Int, y: Int -> 2 * y + x }
+val g = { x: Int, y: Int -> y - x }
+```
+................
+```Haskell
+p t = (t, (-2) * t)
+```
+```kotlin:ank:playground
+import arrow.core.toT
+
+val p = { t: Int -> t toT (-2) * t }
+```
+................
+```Haskell
+f . p' = g . p'
+```
+```kotlin
+f compose `p'` == g compose `p'`
+```
+................
+```Haskell
+p'() = (0, 0)
+```
+```kotlin:ank:playground
+import arrow.core.toT
+
+fun `p'`() = 0 toT 0
+```
+................
+```Haskell
+h () = 0
+```
+```kotlin:ank:plaground
+fun h() = 0
+```
+................
+```Haskell
+f :: a -> b
+g :: c -> b
+```
+```kotlin
+val f: (A) -> B
+val g: (C) -> B
+```
+................
+```Haskell
+p :: d -> a
+q :: d -> c
+r :: d -> b
+```
+```kotlin
+val p: (D) -> A
+val q: (D) -> C
+val r: (D) -> B
+```
+................
+```Haskell
+g . q = f . q
+```
+```kotlin
+g compose q == f compose q
+```
+................
+```Haskell
+f x = 1.23
+```
+```kotlin
+val f: (A) -> Double = { _: A -> 1.23 }
+```
+................
+```Haskell
+twice f x = f (f x)
+```
+```kotlin
+fun <A> twice() = { f: (A) -> A, x: A -> f(f(x)) }
+```
+................
+```Haskell
+f       :: t0
+x       :: t1
+f x     :: t2
+f (f x) :: t3
+```
+```kotlin
+val f        : t0
+val x        : t1
+val `(f x)`  : t2
+val `f (f x)`: t3
+```
+................
+```Haskell
+twice :: t0 -> t1 -> t3
+```
+```kotlin
+val twice: (t0) -> (t1) -> t3
+```
+................
+```Haskell
+t0 = t1 -> t2 -- because f is applied to x
+t0 = t2 -> t3 -- because f is applied to (f x)
+```
+```kotlin
+typealias t0 = t1 -> t2 // because f is applied to x
+typealias t0 = t2 -> t3 // because f is applied to (f x)
+```
+................
+```Haskell
+t1 = t2 = t3 = Int
+twice :: (Int -> Int) -> Int -> Int
+```
+```kotlin
+fun twice(): ((Int) -> Int, Int) -> Int
+```
+................
+```Haskell
+twice :: (t -> t) -> t -> t
+```
+```kotlin
+fun <A> twice(): ((A) -> A, A) -> A
+```
+................
+```Haskell
+newtype ToString a = ToString (a -> String)
+instance Contravariant ToString where 
+    contramap f (ToString g) = ToString (g . f)
+```
+```kotlin:ank:playground
+import arrow.Kind
+import arrow.extension
+import arrow.typeclasses.Contravariant
+
+@extension
+interface ToStringContravariant : Contravariant<ForToString> {
+  override fun <A, B> Kind<ForToString, A>.contramap(f: (B) -> A): Kind<ForToString, B> =
+    ToString { this.fix().f(f(it)) }
+}
+```
+................
+```Haskell
+ToString (Either b c) ~ (b -> String, c -> String)
+```
+```kotlin
+(ToString<Either<A, B>>) ~ ((A) -> String, (B) -> String)
```

Comment:
I believe this is a `FunctionK`

Reply:
I am not sure how I should write it in kotlin.

Because it's not a variable nor a function per se

Reply:
IIRC `~` is a natural transformation (a function that goes from a wrapper to a different wrapper), but I don't know how it interacts with multiple elements. I do know that natural transformation in Arrow is `FunctionK`...so maybe there's something there?

Reply:
I'll look into it, maybe at the end of the week we have a finished PR

Reply:
We can merge it now and fix it later. That's okay with me too.

Reply:
No hurry.  If it does actually take more time, then it will. 

Arrow is lacking a bit more examples on how to work with FunktionK. Let me try it maybe I can come up with more examples. So far, I only used FunktionK.id()

Reply:
This was the response from the community: 
```
cvlad   [26 minutes ago]
The Natural Transformation operator is generally `~>` and has the form `type (~>) f g = forall x. f x -> g x`, so it's not that. `~` is generally used for type equality -- meaning types are the same thing. Informally, it can be used for isomorphic types.
```
```
Imran S   [15 minutes ago]
so it somehow models type aliases
```
```
cvlad   [9 minutes ago]
Well, in Haskell you can write something like:
```f :: (Functor f, f ~ g) => (a -> b) -> g a -> f b
f = fmap```
And it will work. Note that `f` is `Functor`, but since we added `f ~ g`, then `g` is the same as `f` -- a `Functor`.
```
What do you think is better and more understandable. 

```suggestion
typealias ToStringEither<A, B> = ToString<Either<A, B>>

typealias TupleToString<A, B> = Tuple2<(A) -> String, (B) -> String>

ToStringEither == TupleToString
```
Or should I go for an Functor implementation, as he did

Reply:
hmmmmmm this one isn't supported by us. The alias is probably better.

Reply:
I'll write a small issue, just a reminder, if we support something like this in the future

Merge if everything looks good to you.